cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java

```diff
@@ -207,48 +258,49 @@ public CloseableIterator<StructuredRow> scan(Range keyRange, int limit, Field<?>
     if (!schema.isIndexColumn(filterIndex.getName())) {
       throw new InvalidFieldException(schema.getTableId(), filterIndex.getName(), "is not an indexed column");
     }
-    ScannerIterator scannerIterator = new ScannerIterator(getScanner(keyRange), schema);
-    FilterByIndexIterator filterByIndexIterator = new FilterByIndexIterator(scannerIterator, filterIndex, schema);
+    FilterByFieldIterator filterByIndexIterator =
+      new FilterByFieldIterator(getFilterByRangeIterator(keyRange), filterIndex, schema);
     return new LimitIterator(Collections.singleton(filterByIndexIterator).iterator(), limit);
   }
 
+  /*
+   * Scan the nosql DB with provided ranges, find out the most outbound range, scan and filter by ranges,
+   * the result should already be sorted by primary keys.
+   * */
   @Override
   public CloseableIterator<StructuredRow> multiScan(Collection<Range> keyRanges,
                                                     int limit) throws InvalidFieldException, IOException {
-    // Sort the scan keys by the start key and merge overlapping ranges.
-    Deque<ImmutablePair<byte[], byte[]>> scanKeys = new LinkedList<>();
-    keyRanges.stream()
+
+    if (keyRanges.isEmpty()) {
+      return CloseableIterator.empty();
+    }
+
+    // Call scan single range when there's only range
+    if (keyRanges.size() == 1) {
+      return scan(keyRanges.iterator().next(), limit);
+    }
+
+    FilterByRangeIterator filterIterator =
+      new FilterByRangeIterator(getOutboundRangeIterator(keyRanges), keyRanges);
+    return new LimitIterator(filterIterator, limit);
+  }
+
+  /*
+   * Scan the nosql DB with the most outbound range, result should be sorted
+   * */
+  private ScannerIterator getOutboundRangeIterator(Collection<Range> keyRanges) {
+    List<Range> prefixRanges = keyRanges.stream().map(this::getLongestPrefixRange).collect(Collectors.toList());
+    PriorityQueue<byte[]> begins = new PriorityQueue<>(Bytes::compareTo);
+    PriorityQueue<byte[]> ends = new PriorityQueue<>((o1, o2) -> Bytes.compareTo(o2, o1));
```

Comment:
I don't quite understand if this change matches the original implementation, original implementation is to get the overlap and scan on each non-overlapped range. Is this changing it to scan on the most inbound and outbound range? Do we know the performance impact?

Reply:
The original implementation is based on assumption that the given keys prefix range is prefixed, without holes. Now we drop this assumption and need to find the most outbound range, scan once and filter using OR logic with the given ranges. The performance won't be too bad since the existing use cases mostly sack with a fixed prefix range and it's only local sandbox using nosql db.

Reply:
Reasoning here as well: https://github.com/cdapio/cdap/pull/14605#issue-1401576442